src/Compilers/Core/Portable/Operations/OperationFactory.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Semantics
+{
+    internal static class OperationFactory
+    {
+        public static IVariableDeclaration CreateVariableDeclaration(ILocalSymbol variable, IOperation initialValue, SyntaxNode syntax)
+        {
+            return CreateVariableDeclaration(ImmutableArray.Create(variable), initialValue, syntax);
+        }
+
+        public static VariableDeclaration CreateVariableDeclaration(ImmutableArray<ILocalSymbol> variables, IOperation initialValue, SyntaxNode syntax)
+        {
+            return new VariableDeclaration(
+                variables,
+                initialValue,
+                variables.IsDefaultOrEmpty || (initialValue != null && initialValue.IsInvalid),
+                syntax,
+                type: null,
+                constantValue: default(Optional<object>));
+        }
+
+        public static IConditionalChoiceExpression CreateConditionalChoiceExpression(IOperation condition, IOperation ifTrue, IOperation ifFalse, ITypeSymbol resultType, SyntaxNode syntax)
+        {
+            var isInvalid = (condition == null || condition.IsInvalid || ifTrue == null || ifTrue.IsInvalid || ifFalse == null || ifFalse.IsInvalid || resultType == null);
+
+            return new ConditionalChoiceExpression(
+                condition,
+                ifTrue,
+                ifFalse,
+                isInvalid,
+                syntax,
+                resultType,
+                default(Optional<object>));
+        }
+
+        public static IExpressionStatement CreateAssignmentExpressionStatement(IOperation target, IOperation value, SyntaxNode syntax)
+        {
+            var isInvalid = target == null || target.IsInvalid || value == null || value.IsInvalid;
+            var expression = new AssignmentExpression(target, value, isInvalid, syntax, target.Type, default(Optional<object>));
+            return new ExpressionStatement(expression, expression.IsInvalid, syntax, type: null, constantValue: default(Optional<object>));
+        }
+
+        public static IExpressionStatement CreateCompoundAssignmentExpressionStatement(
+            IOperation target, IOperation value, BinaryOperationKind binaryOperationKind, IMethodSymbol operatorMethod, SyntaxNode syntax)
+        {
+            var isInvalid = target == null || target.IsInvalid || value == null || value.IsInvalid;
```

Comment:
Should you be checking if binaryOperationKind is BinaryOperationKind.Invalid here?